From 2ca361efaa7ad8735140676e2a7e1a3401347117 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Fri, 25 Aug 2017 14:37:20 +0200
Subject: [PATCH 09/14] mmc/sdhci: Add card detect method

Detect whether a card is detected. This prevents probing problems
when no card is in the reader. The board apparently reports "present"
for the non-removable device as well, so we can get away with that.

With "non-removable" in the devicetree there is no card detect. Pass
this information on as a quirk and use it in the sdhci driver.
---
 drivers/mmc/sdhci.c      | 13 +++++++++++++
 drivers/mmc/zynq_sdhci.c |  3 +++
 2 files changed, 16 insertions(+)

diff --git a/drivers/mmc/sdhci.c b/drivers/mmc/sdhci.c
index a91a31d..0218e5b 100644
--- a/drivers/mmc/sdhci.c
+++ b/drivers/mmc/sdhci.c
@@ -687,6 +687,17 @@ static int sdhci_init(struct mmc *mmc)
 	return 0;
 }
 
+static int sdhci_get_cd(struct udevice *dev)
+{
+	struct mmc *mmc = mmc_get_mmc_dev(dev);
+	struct sdhci_host *host = mmc->priv;
+
+	if (host->quirks & SDHCI_QUIRK_NO_CD)
+		return -ENOSYS;
+
+	return !!(sdhci_readl(host, SDHCI_PRESENT_STATE) & SDHCI_CARD_PRESENT);
+}
+
 #ifdef CONFIG_DM_MMC_OPS
 int sdhci_probe(struct udevice *dev)
 {
@@ -701,6 +712,7 @@ const struct dm_mmc_ops sdhci_ops = {
 	.set_voltage	= sdhci_set_voltage,
 	.set_uhs	= sdhci_set_uhs,
 	.execute_tuning	= sdhci_execute_tuning,
+	.get_cd         = sdhci_get_cd,
 };
 #else
 static const struct mmc_ops sdhci_ops = {
@@ -710,6 +722,7 @@ static const struct mmc_ops sdhci_ops = {
 	.set_voltage	= sdhci_set_voltage,
 	.set_uhs	= sdhci_set_uhs,
 	.execute_tuning	= sdhci_execute_tuning,
+	.get_cd         = sdhci_get_cd,
 };
 #endif
 
diff --git a/drivers/mmc/zynq_sdhci.c b/drivers/mmc/zynq_sdhci.c
index 454d3cb..ffb28b5 100644
--- a/drivers/mmc/zynq_sdhci.c
+++ b/drivers/mmc/zynq_sdhci.c
@@ -184,6 +184,9 @@ static int arasan_sdhci_probe(struct udevice *dev)
 		       SDHCI_QUIRK_BROKEN_R1B |
 		       SDHCI_QUIRK_USE_ACMD12;
 
+	if (fdtdec_get_bool(gd->fdt_blob, dev->of_offset, "non-removable"))
+		host->quirks |= SDHCI_QUIRK_NO_CD;
+
 #ifdef CONFIG_ZYNQ_HISPD_BROKEN
 	host->quirks |= SDHCI_QUIRK_NO_HISPD_BIT;
 #endif
-- 
1.9.1

